### ISSUE TITLE
Generic types.Event overhead

### ISSUE THREAD

**Issue by NDStrahilevitz** [aquasecurity/tracee#1431]
# The issue itself

In tracee-rules, the existing `Event` type is simply the generic `interface{}`. This has implications on the performance of tracee-rules and it's cpu time demand. During the performance analysis that @mtcherni95 and I are performing, we've noticed that convT2E can take from 10-20% of overall cpu time. 
In our case, since we're consuming tracee-rules and tracee-ebpf as libraries (single binary), that has the further effect of taking cpu time from tracee-ebpf, thus increasing the chance of event loss.

The simplest way to remove this overhead would be to replace `external.Event` with`types.Event` everywhere. This eliminates the problem entirely, and in our performance tests introduced performance gains (shown in the performance analysis below). However it did mean breaking the Signature interface API, which obviously means this change can't be taken too lightly.

If it's agreed to pursue this agenda I see two reasonable ways to go about it:

1. As said before - replacing `types.Event` with `external.Event` everywhere. This means loss of generality with the gain of efficiency, however since practically all signatures type check for `external.Event` anyway, and considering there is logic in tracee-rules tightly dependent on `external.Event` the loss of generality isn't very "real". This way can also be done as a first step towards a next more general solution
2. Immediately reimplementing the generality concept. In discussions I've had on this topic the idea of adding an `EventHeader` type was raised, which we could immediately supply along with tracee-ebpf events, and enrich other event sources if we want to add them to tracee-rules (unlike now where we would have to hard code their support). Signatures could then support events according to a header and not through type cast checking.
3. There are probably more ways to go about it, I'd like to use this issue to discuss them :)

This is a big issue to tackle but I think it's worth it in terms of performance and in establishing a better implementation of consuming events from various sources.

# Performance analysis:

There are 3 images showing the point of the optimization (running on the previously mentioned single binary tracee).

1. Flame graph with cpu usage of tracee.processEvents (buffered signature channels(see #1429) and no convT2E optimization).
![image](https://user-images.githubusercontent.com/22661609/152170479-a5241c94-f269-42f1-8cf2-b706ace82626.png)
2. Flame graph with cpu usage of tracee.processEvents (buffered signature channels(see #1429) and including convT2E optimization). As can be seen there was an ~8% cpu usage increase for tracee-ebpf.
![image](https://user-images.githubusercontent.com/22661609/152170555-79d06512-af40-4bc8-9398-8d289514f647.png)
3. Difference of the runtime.convT2E cpu usage from a pprof diff graph
![image](https://user-images.githubusercontent.com/22661609/152169804-6d2ea882-8ee6-4254-8dc3-fcf3f98e6f83.png)

And the pprofs themselves:
[1000buffer.pb.gz](https://github.com/aquasecurity/tracee/files/7987163/1000buffer.pb.gz)
[convt2e-1000-buffer.pb.gz](https://github.com/aquasecurity/tracee/files/7987165/convt2e-1000-buffer.pb.gz)

**Comment by danielpacak** [comment]
I agree that Go `interface{}` used to represent generic event and Go type system along with Go type casing used to dispatch those events is problematic.

In a nutshell, in Messaging Systems a common approach would be to split an event into event metadata (aka headers) and event payload by implementing [Canonical Data Model] and [Format Indicator] patterns. This would allow us, for example, to tell tracee `external.Event`s from Kubernetes audit events or any other event source. Having event metadata or event headers would allow us to [dispatch events](https://www.enterpriseintegrationpatterns.com/MessageDispatcher.html) to tracee-rules engine or signatures without always parsing the whole event payload.

Another thing that we may check and possibly use is [CloudEvents](https://cloudevents.io) specification.

[Canonical Data Model]: https://www.enterpriseintegrationpatterns.com/patterns/messaging/CanonicalDataModel.html
[Format Indicator]: https://www.enterpriseintegrationpatterns.com/patterns/messaging/FormatIndicator.html

**Comment by NDStrahilevitz** [comment]
Thanks for the feedback @danielpacak. I think using CloudEvents is a good direction for a solution so i'll be going ahead with it for now. 
Regardless, moving from `interface {}` is likely going to break backwards compatibility so we need to be prepared for that in terms of signatures (I think this is more for @AsafEitani)

**Comment by AsafEitani** [comment]
I just faced the issue myself but from another perspective. After we tried closing #1312 by separating `types` into a different module we faced a similar issue with `external`. We agreed that we will combine the two into a single module.
While I was working today to combine those two packages I came across the redeclaration of `Event`.

If we do combine those (and I believe we should) I think that we will have to break backward compatibility. Essentially by combining the two packages into a single package we are forced to go with the first option @NDStrahilevitz suggested.

@danielpacak Any thoughts?

**Comment by NDStrahilevitz** [comment]
> I just faced the issue myself but from another perspective. After we tried closing #1312 by separating `types` into a different module we faced a similar issue with `external`. We agreed that we will combine the two into a single module. While I was working today to combine those two packages I came across the redeclaration of `Event`.
> 
> If we do combine those (and I believe we should) I think that we will have to break backward compatibility. Essentially by combining the two packages into a single package we are forced to go with the first option @NDStrahilevitz suggested.
> 
> @danielpacak Any thoughts?

It would be really beneficial to me to have them merged before a PR solving this, so if there isn't any more discussion to be had regarding merging them I'd like to upload one doing that before having one solving the event issue here.
Is there a relevant issue regarding merging them?

**Comment by AsafEitani** [comment]
> > I just faced the issue myself but from another perspective. After we tried closing #1312 by separating `types` into a different module we faced a similar issue with `external`. We agreed that we will combine the two into a single module. While I was working today to combine those two packages I came across the redeclaration of `Event`.
> > If we do combine those (and I believe we should) I think that we will have to break backward compatibility. Essentially by combining the two packages into a single package we are forced to go with the first option @NDStrahilevitz suggested.
> > @danielpacak Any thoughts?
> 
> It would be really beneficial to me to have them merged before a PR solving this, so if there isn't any more discussion to be had regarding merging them I'd like to upload one doing that before having one solving the event issue here. Is there a relevant issue regarding merging them?

I would wait for @danielpacak to respond. If the merging offer will be accepted I will do a PR on merging the two packages.

**Comment by NDStrahilevitz** [comment]
Updating from a sync I've had with @AsafEitani, we've decided currently that the `signature.OnEvent` API will change to `OnEvent(payload []byte, contentType string)` thus allowing to pass a generic payload and a "context" by which to decode.
The other option that was considered was passing the entire CloudEvent but that means restricting the envelope type we will support, this solution will allow us to support other message types in the future.

**Comment by NDStrahilevitz** [comment]
@danielpacak I was wondering if we're going the CloudEvents route, if findings should also be reported through a CloudEvent, that way we could even republish them back to tracee-rules under appropriate conditions or integrate them with other CloudEvents supporting tools, let me know what you think.

**Comment by mtcherni95** [comment]
> Updating from a sync I've had with @AsafEitani, we've decided currently that the `signature.OnEvent` API will change to `OnEvent(payload []byte, contentType string)` thus allowing to pass a generic payload and a "context" by which to decode. The other option that was considered was passing the entire CloudEvent but that means restricting the envelope type we will support, this solution will allow us to support other message types in the future.

quick question. Why passing a string for contentType? I guess that according to the type the handling will be different. Why not just passing an integer (or even enum)?

**Comment by NDStrahilevitz** [comment]
> > Updating from a sync I've had with @AsafEitani, we've decided currently that the `signature.OnEvent` API will change to `OnEvent(payload []byte, contentType string)` thus allowing to pass a generic payload and a "context" by which to decode. The other option that was considered was passing the entire CloudEvent but that means restricting the envelope type we will support, this solution will allow us to support other message types in the future.
> 
> quick question. Why passing a string for contentType? I guess that according to the type the handling will be different. Why not just passing an integer (or even enum)?

As I see it there's a tradeoff in both approaches, with a string we get more flexibility (but room for errors) and with an enum
 we would need to explicitly declare which event types tracee-rules support before we could add a signature supporting such an event, so we gain safety but reduce future flexibility. @AsafEitani @danielpacak your thoughts would be appreciated.

**Comment by AsafEitani** [comment]
> > Updating from a sync I've had with @AsafEitani, we've decided currently that the `signature.OnEvent` API will change to `OnEvent(payload []byte, contentType string)` thus allowing to pass a generic payload and a "context" by which to decode. The other option that was considered was passing the entire CloudEvent but that means restricting the envelope type we will support, this solution will allow us to support other message types in the future.
> 
> quick question. Why passing a string for contentType? I guess that according to the type the handling will be different. Why not just passing an integer (or even enum)?

I agree with @NDStrahilevitz, if someone wants to use tracee-rules with a new source he will be able to. With enums he will need to modify tracee-rules source code.
This comes with a price of comparing strings twice in each event - first to check to which signature to send this event and second in the signature itself, in case that multiple sources are supported in this signature.

Enums would be more efficient but would make tracee-rules less generic for unsupported sources.

**Comment by danielpacak** [comment]
I think that the first step might be to use generic representation of event in tracee without introducing enums, string constants, dependency on 3rd party libraries like CloudEvents, and to avoid proprietary message representations.

One way to do that is separate message payload from message metadata, which is a common approach in messaging system. For example, Java and Spring is doing it for ages with its [`Message<T>`](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/messaging/Message.html) interface. In Go the same can be as simple as the following structure:

```go
package types

// Event is a generic message representation with headers and body.
type Event struct {
	Headers map[string]string
	Payload []byte
}
```

Once we have that we can define mandatory headers for tracee-rules. For example, add `ContentType` header. Other popular and well defined headers are CorrelationID, SequenceID, etc.

With that we can change the signature of the OnEvent method to `OnEvent(event types.Event)`.

**Comment by NDStrahilevitz** [comment]
The overhead was solved in #1482.

I think implications of the change should be discussed in another issue (or issues) since it probably allows to do new things with `protocol.Event`
